### PR TITLE
Refactor Communication framework

### LINKED ISSUE #460: Refactor GRPC communication capsulation

### LINKED ISSUE #682: Refactor grpc server
### Feature request:
- refactor grpc server.

### PR DESCRIPTION
**Pull request by 244372610**
## Description
* Provide interceptor for service node migration
* Provide ElasticClient to obtain peertasks
* Use a more native way to complete GRPC calls
* Replace dferrors with status.Error

## Related Issue
fix #460 
fix #682 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [ ] Code compiles correctly.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: pkg/rpc/scheduler/server/server.go [hunk 1]
```diff
@@ -32,7 +36,7 @@ type SchedulerServer interface {
 	// RegisterPeerTask registers a peer into one task.
 	RegisterPeerTask(context.Context, *scheduler.PeerTaskRequest) (*scheduler.RegisterResult, error)
 	// ReportPieceResult reports piece results and receives peer packets.
-	ReportPieceResult(scheduler.Scheduler_ReportPieceResultServer) error
+	ReportPieceResult(context.Context, scheduler.Scheduler_ReportPieceResultServer) error
```
Comment (gaius-qi) [inline pkg/rpc/scheduler/server/server.go:39]: Designed according to the native `GRPC` function.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/rpcserver/rpcserver.go [hunk 2]
```diff
@@ -21,38 +21,53 @@ import (
...
-	svr.Server = schedulerserver.New(svr, opts...)
+	grpcServer := grpc.NewServer(append(rpc.DefaultServerOptions, opts...)...)
+	scheduler.RegisterSchedulerServer(grpcServer, svr)
+	svr.Server = grpcServer
 	return svr
 }
 
-func (s *Server) RegisterPeerTask(ctx context.Context, req *scheduler.PeerTaskRequest) (*scheduler.RegisterResult, error) {
+func (s *Server) RegisterPeerTask(ctx context.Context, req *scheduler.PeerTaskRequest) (resp *scheduler.RegisterResult, err error) {
+	metrics.RegisterPeerTaskCount.Inc()
```
Comment (gaius-qi) [inline scheduler/rpcserver/rpcserver.go:58]: Mtrics is put back in the proxy and reply proxy code.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/rpc/balancer.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+	// If resolver state contains no addresses, return an error so ClientConn
+	// will trigger re-resolve. Also records this as a resolver error, so when
+	// the overall state turns transient failure, the error message will have
+	// the zero address information.
+	if len(s.ResolverState.Addresses) == 0 {
+		b.ResolverError(fmt.Errorf("produced zero addresses"))
+		return balancer.ErrBadResolverState
+	}
+
+	// As we want to do the connection management ourselves, we don't set up connection here.
```
Comment (zzy987) [inline pkg/rpc/balancer.go:161]: Are the comments here out of date?

Comment (244372610) [reply pkg/rpc/balancer.go:161]: Comments still apply

File: pkg/rpc/cdnsystem/testdata/seed_piece_info.json [hunk 4]
```diff
@@ -0,0 +1,171 @@
+{
+  "aaaaaaaaaaaaaaa": [
```
Comment (gaius-qi) [inline pkg/rpc/cdnsystem/testdata/seed_piece_info.json:2]: Rename `foo` or `bar`.

File: pkg/rpc/cdnsystem/testdata/seed_piece_info.json [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        "piece_md5": "ddd",
+        "piece_offset": 500,
+        "piece_style": 1
+      },
+      "done": true,
+      "content_length": 501,
+      "total_piece_count": 6
+    }
+  ],
+  "bbbbbbbbbbbbbbb": [
```
Comment (gaius-qi) [inline pkg/rpc/cdnsystem/testdata/seed_piece_info.json:94]: Rename `foo` or `bar`.

File: pkg/rpc/dfdaemon/testdata/pieces.json [hunk 6]
```diff
@@ -0,0 +1,111 @@
+{
+  "aaaaaaaaaaaaaaa": {
```
Comment (gaius-qi) [inline pkg/rpc/dfdaemon/testdata/pieces.json:2]: Rename `foo` or `bar`.

File: pkg/rpc/dfdaemon/testdata/pieces.json [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        "piece_md5": "ddd",
+        "piece_offset": 500,
+        "piece_style": 1
+      }
+    ],
+    "piece_md5_sign": "aaaaaaaaaaaaaaa_sign",
+    "content_length": 501,
+    "total_piece": 6
+  },
+  "bbbbbbbbbbbbbbb": {
```
Comment (gaius-qi) [inline pkg/rpc/dfdaemon/testdata/pieces.json:57]: Rename `foo` or `bar`.

File: pkg/rpc/dfdaemon/testdata/pieces.json [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      {
+        "piece_num": 5,
+        "range_start": 500,
+        "range_size": 1,
+        "piece_md5": "ddd",
+        "piece_offset": 500,
+        "piece_style": 1
+      }
+    ],
+    "piece_md5_sign": "aaaaaaaaaaaaaaa_sign",
```
Comment (gaius-qi) [inline pkg/rpc/dfdaemon/testdata/pieces.json:53]: Rename `foo` or `bar`.

File: pkg/rpc/dfdaemon/testdata/pieces.json [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+          "piece_md5": "ddd",
+          "piece_offset": 4000,
+          "piece_style": 1
+        },
+        "done": true,
+        "content_length": 4500,
+        "total_piece_count": 5
+      }
+    ],
+    "piece_md5_sign": "bbbbbbbbbbbbbb_sign",
```
Comment (gaius-qi) [inline pkg/rpc/dfdaemon/testdata/pieces.json:107]: Rename `foo` or `bar`.

File: pkg/rpc/scheduler/client/mocks/mock_client.go [hunk 10]
```diff
@@ -1,16 +1,15 @@
 // Code generated by MockGen. DO NOT EDIT.
```
Comment (gaius-qi) [inline pkg/rpc/scheduler/client/mocks/mock_client.go:1]: Change file name to `client_mock.go`.

File: pkg/rpc/scheduler/mocks/mock_scheduler.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (gaius-qi) [inline pkg/rpc/scheduler/mocks/mock_scheduler.go:1]: Rename file name to `scheduler_mock.go`.

File: scheduler/resource/cdn_mock.go [hunk 12]
```diff
@@ -11,7 +11,6 @@ import (
 	dfnet "d7y.io/dragonfly/v2/internal/dfnet"
```
Comment (gaius-qi) [inline scheduler/resource/cdn_mock.go:11]: Revert this file.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

**Pull request closed without merging**